src/librbd/Utils.cc

```diff
@@ -242,5 +247,152 @@ int get_config_key(librados::Rados& rados, const std::string& uri,
   return 0;
 }
 
+const std::vector<std::string> UNIQUE_PEER_CONFIG_KEYS {
+  {"monmap", "mon_host", "mon_dns_srv_name", "key", "keyfile", "keyring"}};
+
+int init_rados(CephContext *cct,
+               const std::string &cluster_name,
+               const std::string &client_name,
+               const std::string &mon_host,
+               const std::string &key,
+               const std::string &description,
+               std::vector<const char*>& args,
+               librados::Rados **rados_ref,
+               bool strip_cluster_overrides,
+               const std::string &admin_socket) {
+  // NOTE: manually bootstrap a CephContext here instead of via
+  // the librados API to avoid mixing global singletons between
+  // the librados shared library and the daemon
+  // TODO: eliminate intermingling of global singletons within Ceph APIs
+  CephInitParameters iparams(CEPH_ENTITY_TYPE_CLIENT);
+  if (client_name.empty() || !iparams.name.from_str(client_name)) {
+    lderr(cct) << "error initializing cluster handle for " << description << dendl;
+    return -EINVAL;
+  }
+
+  CephContext *new_cct = common_preinit(iparams, CODE_ENVIRONMENT_LIBRARY,
+                                    CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS);
+  new_cct->_conf->cluster = cluster_name;
+
+  // librados::Rados::conf_read_file
+  int r = new_cct->_conf.parse_config_files(nullptr, nullptr, 0);
+  if (r < 0 && r != -ENOENT) {
+    // do not treat this as fatal, it might still be able to connect
+    lderr(cct) << "could not read ceph conf for " << description << ": "
+	             << cpp_strerror(r) << dendl;
+  }
+
+  // preserve cluster-specific config settings before applying environment/cli
+  // overrides
+  std::map<std::string, std::string> config_values;
+  if (strip_cluster_overrides) {
+    // remote peer connections shouldn't apply cluster-specific
+    // configuration settings
+    for (auto& key : UNIQUE_PEER_CONFIG_KEYS) {
+      config_values[key] = new_cct->_conf.get_val<std::string>(key);
+    }
+  }
+
+  new_cct->_conf.parse_env(new_cct->get_module_type());
+
+  // librados::Rados::conf_parse_env
+  std::vector<const char*> env_args;
+  r = new_cct->_conf.parse_argv(env_args);
+  if (r < 0) {
+    lderr(cct) << "could not parse environment for " << description << ":"
+               << cpp_strerror(r) << dendl;
+    new_cct->put();
```

Comment:
nit: we could use `scope_guard` from `include/scope_guard.h` to do `new_cct->put()` on return automatically. But it is up to you.